lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,63 +1175,202 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
+  switch (Value) {
+  case AutolinkKind::LLVMDependentLibraries:
+    return "llvm.dependent-libraries";
+  case AutolinkKind::LLVMLinkerOptions:
+  case AutolinkKind::SwiftAutoLinkExtract:
+    return "llvm.linker.options";
+  }
 
-  // Remove duplicates.
-  llvm::SmallPtrSet<llvm::MDNode *, 4> knownAutolinkEntries;
-  AutolinkEntries.erase(std::remove_if(AutolinkEntries.begin(),
-                                       AutolinkEntries.end(),
-                                       [&](llvm::MDNode *entry) -> bool {
-                                         return !knownAutolinkEntries.insert(
-                                                   entry).second;
-                                       }),
-                        AutolinkEntries.end());
-
-  const bool AutolinkExtractRequired =
-      (TargetInfo.OutputObjectFormat == llvm::Triple::ELF && !Triple.isPS4()) ||
-      TargetInfo.OutputObjectFormat == llvm::Triple::Wasm ||
-      Triple.isOSCygMing();
-
-  if (!AutolinkExtractRequired) {
+  llvm_unreachable("Unhandled AutolinkKind in switch.");
+}
+
+void AutolinkKind::collectEntriesFromLibraries(
+    llvm::SmallVector<llvm::MDNode *, 4> &Entries,
+    const SmallVector<LinkLibrary, 32> &AutolinkEntries, IRGenModule &IGM) {
+  llvm::LLVMContext &ctx = IGM.getLLVMContext();
+
+  switch (Value) {
+  case AutolinkKind::LLVMLinkerOptions:
+  case AutolinkKind::SwiftAutoLinkExtract: {
+    // On platforms that support autolinking, continue to use the metadata.
+    for (LinkLibrary linkLib : AutolinkEntries) {
+      switch (linkLib.getKind()) {
+      case LibraryKind::Library: {
+        llvm::SmallString<32> opt =
+            getTargetDependentLibraryOption(IGM.Triple, linkLib.getName());
+        Entries.push_back(
+            llvm::MDNode::get(ctx, llvm::MDString::get(ctx, opt)));
+        continue;
+      }
+      case LibraryKind::Framework: {
+        llvm::Metadata *args[] = {llvm::MDString::get(ctx, "-framework"),
+                                  llvm::MDString::get(ctx, linkLib.getName())};
+        Entries.push_back(llvm::MDNode::get(ctx, args));
+        continue;
+      }
+      }
+      llvm_unreachable("Unhandled LibraryKind in switch.");
+    }
+    return;
+  }
+  case AutolinkKind::LLVMDependentLibraries: {
+    for (LinkLibrary linkLib : AutolinkEntries) {
+      switch (linkLib.getKind()) {
+      case LibraryKind::Library: {
+        Entries.push_back(llvm::MDNode::get(
+            ctx, llvm::MDString::get(ctx, linkLib.getName())));
+        continue;
+      }
+      case LibraryKind::Framework: {
+        llvm_unreachable(
+            "llvm.dependent-libraries doesn't support framework dependency");
+        return;
+      }
+      }
+      llvm_unreachable("Unhandled LibraryKind in switch.");
+    }
+    return;
+  }
+  }
+  llvm_unreachable("Unhandled AutolinkKind in switch.");
+}
+
+void AutolinkKind::writeEntries(
+    const llvm::SmallVector<llvm::MDNode *, 4> &Entries,
```

Comment:
If you aren't going to actually write to this, I would change it to be an ArrayRef<llvm::MDNode *>. This is more general and doesn't have the small vector size in the type.

If one needed to append to the array, the way people do it is that they use:

```
SmallVectorImpl<llvm::MDNode *> &Entries
```

that way we avoid encoding the 4. That being said I think this is an immutable use here, so this is just an FYI.